homeassistant/components/xiaomi_miio/fan.py

```diff
@@ -674,3 +696,192 @@ async def async_reset_filter(self):
+            | SUPPORT_PRESET_MODE
+            | SUPPORT_DIRECTION
+        )
+        self._preset_mode = None
+        self._oscillating = None
+        self._percentage = None
+
+    @property
+    def preset_mode(self):
+        """Get the active preset mode."""
+        return ATTR_MODE_NATURE if self._nature_mode else ATTR_MODE_NORMAL
+
+    @property
+    def percentage(self):
+        """Return the current speed as a percentage."""
+        return self._percentage
+
+    @property
+    def oscillating(self):
+        """Return whether or not the fan is currently oscillating."""
+        return self._oscillating
+
+    @callback
+    def _handle_coordinator_update(self):
+        """Fetch state from the device."""
+        self._available = True
+        self._state = self.coordinator.data.is_on
+        self._oscillating = self.coordinator.data.oscillate
+        self._nature_mode = self.coordinator.data.natural_speed != 0
+        if self.coordinator.data.is_on:
+            if self._nature_mode:
+                self._percentage = self.coordinator.data.natural_speed
+            else:
+                self._percentage = self.coordinator.data.direct_speed
+        else:
+            self._percentage = 0
+
+        self.async_write_ha_state()
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set the preset mode of the fan."""
+        if preset_mode not in self.preset_modes:
+            _LOGGER.warning("'%s'is not a valid preset mode", preset_mode)
+            return
+
+        if preset_mode == ATTR_MODE_NATURE:
+            await self._try_command(
+                "Setting fan speed percentage of the miio device failed.",
+                self._device.set_natural_speed,
+                self._percentage,
+            )
+        else:
+            await self._try_command(
+                "Setting fan speed percentage of the miio device failed.",
+                self._device.set_direct_speed,
+                self._percentage,
+            )
+
+        self._preset_mode = preset_mode
+        self.async_write_ha_state()
+
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the percentage of the fan."""
+        if percentage == 0:
+            await self.async_turn_off()
+            self._percentage = 0
+            self.async_write_ha_state()
+            return
+
+        if self._nature_mode:
+            await self._try_command(
+                "Setting fan speed percentage of the miio device failed.",
+                self._device.set_natural_speed,
+                percentage,
+            )
+        else:
+            await self._try_command(
+                "Setting fan speed percentage of the miio device failed.",
+                self._device.set_direct_speed,
+                percentage,
+            )
+        self._percentage = percentage
+
+        if not self.is_on:
+            await self.async_turn_on()
+
+        self.async_write_ha_state()
```

Comment:
```suggestion
        else:
            self.async_write_ha_state()
```

Looks like async_turn_on will write state as well